draft-ietf-quic-http.md

```diff
@@ -667,30 +667,35 @@ this limit are not guaranteed to be accepted.
 
 ### Request Cancellation and Rejection {#request-cancellation}
 
-Clients can cancel requests by resetting and aborting the request stream with an
-error code of H3_REQUEST_CANCELLED ({{http-error-codes}}).  When the client
-aborts reading a response, it indicates that this response is no longer of
-interest. Implementations SHOULD cancel requests by abruptly terminating any
-directions of a stream that are still open.
-
-When the server rejects a request without performing any application processing,
-it SHOULD abort its response stream with the error code H3_REQUEST_REJECTED.
-In this context, "processed" means that some data from the stream was passed to
-some higher layer of software that might have taken some action as a result. The
-client can treat requests rejected by the server as though they had never been
-sent at all, thereby allowing them to be retried later on a new connection.
-Servers MUST NOT use the H3_REQUEST_REJECTED error code for requests which
-were partially or fully processed.  When a server abandons a response after
-partial processing, it SHOULD abort its response stream with the error code
-H3_REQUEST_CANCELLED.
-
-When a client resets a request with the error code H3_REQUEST_CANCELLED, a
-server MAY abruptly terminate the response using the error code
-H3_REQUEST_REJECTED if no processing was performed.  Clients MUST NOT use the
-H3_REQUEST_REJECTED error code, except when a server has requested closure of
-the request stream with this error code.
-
-If a stream is cancelled after receiving a complete response, the client MAY
+Once a request has been sent, it MAY be cancelled by either endpoint.  Clients
+cancel requests if the response is no longer of interest; servers cancel
+requests if they are unable to or choose not to respond.  When possible, it is
+RECOMMENDED that servers send an HTTP response with an appropriate status code
+rather than canceling a request it has already begun processing.
+
+Implementations SHOULD cancel requests by abruptly terminating any
+directions of a stream that are still open.  This means resetting the
+sending parts of streams and aborting reading on receiving parts of streams;
+see Section 2.4 of [QUIC-TRANSPORT].
+
+When the server cancels a request without performing any application processing,
+the request is considered "rejected."  The server SHOULD abort its response
+stream with the error code H3_REQUEST_REJECTED. In this context, "processed"
+means that some data from the stream was passed to some higher layer of software
+that might have taken some action as a result. The client can treat requests
+rejected by the server as though they had never been sent at all, thereby
+allowing them to be retried later on a new connection. Servers MUST NOT use the
```

Comment:
Is there any scenario whereby the request could be retried on the same connection (in which case "on a new connection" would be too restrictive)?  BTW, why would a server reject a request but not close the connection?

Reply:
Good point.  In some ways, this is a holdover from HTTP/2, where it was possible for the client to exceed the server's stream limit, and REJECTED was a mechanism to close those streams without penalizing the client.  A server might still do that under load -- you can't take back stream credit, but you can reject incoming requests and not replace the stream credit.

Reply:
That makes sense.  In that case, a client is allowed to retry the request later on the same connection.